src/v/archival/ntp_archiver_service.cc

```diff
@@ -2010,18 +2010,21 @@ ntp_archiver::get_housekeeping_jobs() {
     return res;
 }
 
-ss::future<std::optional<upload_candidate_with_locks>>
+ss::future<std::pair<
+  std::optional<ssx::semaphore_units>,
```

Comment:
I don't think that you need to carry the semaphore units outside of the call. Once the segment locks are acquired you can dispose the semaphore units. The `upload` call can then acquire the semaphore units separately.

Reply:
Wouldn't that result in the same lock inversion? If we drop the units, the archival loop may take the mutex and attempt to lock the segments and it'd result in the same deadlock, no?

Reply:
I'm concerned about lock order like:

M: takes archiver mutex
M: takes segment locks to find upload candidate
M: drops archiver mutex
S: waits to acquire segment write locks
A: takes archiver mutex
A: waits to acquire segment read locks (blocked behind S)
M: waits to acquire archiver mutex, but everything is waiting on the segment locks that M holds

If we drop and then try to reacquire while still holding the segment locks, it's as if the lock ordering is segment lock then mutex lock, which is problematic, unless we also drop the segment locks before uploading and then take them again before uploading.

Anyway, we can revisit this when there's more time to reorganize some of the archival code